specs/language/conversions.tex

```diff
@@ -0,0 +1,135 @@
+refers to an object that is not of type \texttt{T} and is not an object of a
+type derived from \texttt{T}, the program is ill-formed. If the glvalue refers to
+an object that is uninitialized, the behavior is undefined. Otherwise the
+prvalue is of type \texttt{T}.
+
+\p If the glvalue refers to an array of type \texttt{T}, the prvalue will refer
+to a copy of the array, not memory referred to by the glvalue.
+
+\Sec{Array-to-pointer conversion}{Conv.array}
+
+\p An lvalue or rvalue of type \texttt{T[]} (bounded or unbounded), can be
+converted to a prvalue of type pointer to \texttt{T}.
+[\textit{Note: \acrshort{hlsl} does not support grammar for specifying pointer or
+reference types, however they are used in the type system and must be described
+in language rules.}]
+
+\Sec{Integral conversion}{Conv.iconv}
+
+\p A glvalue of an integer type can be converted to a cxvalue of any other
+non-enumeration integer type. A prvalue of an integer type can be converted to a
+prvalue of any other integer type.
+
+\p If the destination type is unsigned, integer conversion maintains the bit pattern
+of the source value in the destination type truncating or extending the value to
+the destination type.
+
+\p If the destination type is signed, the value is unchanged if the destination
+type can represent the source value. If the destination type cannot represent
+the source value, the result is implementation-defined.
+
+\p If the source type is \texttt{bool}, the values \texttt{true} and
+\texttt{false} are converted to one and zero respectively.
+
+\Sec{Floating point conversion}{Conv.fconv}
+
+\p A glvalue of a floating point type can be converted to a cxvalue of any other
+floating point type. A prvalue of a floating point type can be converted to a
+prvalue of any other floating point type.
+
+\p If the source value can be exactly represented in the destination type, the
+conversion produces the exact representation of the source value. If the source
+value cannot be exactly represented, the conversion to a best-approximation of
+the source value is implementation defined.
+
+\Sec{Floating point-integral conversion}{Conv.fpint}
+
+\p A glvalue of floating point type can be converted to a cxvalue of integer
+type. A prvalue of floating point type can be converted to a prvalue of
+integer type. Conversion of floating point values to integer values truncates by
+discarding the fractional value. The behavior is undefined if the truncated
+value cannot be represented in the destination type.
+
+\p A glvalue of integer type can be converted to a cxvalue of floating point
+type. A prvalue of integer type can be converted to a prvalue of floating
+point type. If the destination type can exactly represent the source value, the
+result is the exact value. If the destination type cannot exactly represent the
+source value, the conversion to a best-approximation of the source value is
+implementation defined.
```

Comment:
Do we need to talk about rounding modes here? I worry that "best-approximation" is very ambiguous otherwise. If we're just going to kick the can and worry about those details later I'd just say "approximation".

Reply:
I think the rounding modes falls under the "implementation defined" aspect and I would expect us to document the DXIL & SPIR-V behaviors separately since they may be different.